core/src/main/java/org/infinispan/stream/impl/AbstractCacheStream.java

```diff
@@ -283,24 +284,23 @@ public void close() {
             boolean localRun = ch.getMembers().contains(localAddress);
             if (localRun) {
                localValue = op.performOperation();
+
+               IntSet ourSegments;
+               if (segmentsToProcess != null) {
+                  ourSegments =  IntSets.mutableFrom(cacheTopology.getLocalReadSegments());
+                  ourSegments.retainAll(segmentsToProcess);
+               } else {
+                  ourSegments = IntSets.from(cacheTopology.getLocalReadSegments());
+               }
                // TODO: we can do this more efficiently - since we drop all results locally
-               if (dm.getReadConsistentHash().equals(ch)) {
-                  IntSet ourSegments;
-                  if (segmentsToProcess != null) {
-                     ourSegments =  IntSets.mutableFrom(ch.getPrimarySegmentsForOwner(localAddress));
-                     ourSegments.retainAll(segmentsToProcess);
-                  } else {
-                     ourSegments = IntSets.from(ch.getPrimarySegmentsForOwner(localAddress));
-                  }
+               LocalizedCacheTopology newCacheTopology = dm.getCacheTopology();
```

Comment:
@wburns I'm not sure about the comment here, does it mean we could do better than drop all the local segments, or does it mean that we could replace the CH equality check with something more efficient?

Reply:
Sorry, in the side-by-side view I was writing my comment just below the comment on line 286 (295 with my changes) :)

I was thinking that would work if the local node didn't lose any segments in the new CH, so we could check directly if `newCacheTopology.getLocalReadSegments()` is a superset of `cacheTopology.getLocalReadSegments()` or not instead of checking all the segments of all the nodes.

I'm not sure I understood your comment, did you have something like this in mind, or were you thinking about when `segmentsToFilter` is set and we don't even need all the segments that were local in the initial CH?

Reply:
This is what I was thinking, but unfortunately using `newCacheTopology.getLocalReadSegments` and `cacheTopology.getLocalReadSegments` is not sufficient. It is entirely possible for the cache topology to change more than once, and I have tests where a segment can move from owned to not owned and back to owned again :( In this case it is possible for an entry to be missed. This is why the LocalStreamManagerImpl registers a listener and tracks the segments across all the topology changes not just the first and most recent.